### PR TITLE
feat: Display more helpful message when encountering Windows long path problem

### PR DESCRIPTION
**Pull request by lucashuy**
#### Which issue(s) does this change fix?
#4031 

#### Why is this change necessary?
This change checks for a `FileNotFoundException`, which can happen if `sam init` is run on Windows without enabling long paths in the registry.

#### How does it address the issue?
Catches the `FileNotFoundException` and displays a more helpful message to Windows users.

```
Cloning from https://github.com/aws/aws-sam-cli-app-templates (process may take a moment)
Error: Failed to modify a local file when cloning app templates. MAX_PATH should be enabled in the Windows registry.
For more details on how to enable MAX_PATH, please visit: <link to documentation>
```

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [ ] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: tests/unit/lib/utils/test_osutils.py [hunk 1]
```diff
@@ -128,3 +128,12 @@ def test_must_copy_if_symlink_fails(self, patched_copy_tree, pathced_os, patched
...
         patched_copy_tree.assert_called_with(source_path, destination_path)
+
+
+class Test_rmtree_callback(TestCase):
+    @patch("samcli.lib.utils.osutils.os")
+    def test_rmtree_callback_raises_file_not_found_exception(self, patched_os):
+        patched_os.chmod.side_effect = FileNotFoundError("File not found")
+
+        with self.assertRaises(FileNotFoundError):
+            rmtree_callback(Mock(), "/some/path", Mock())
```
Comment (sriram-mv) [inline tests/unit/lib/utils/test_osutils.py:139]: nit: I know we are mocking here, does the `/some/path` being UNIX path matter at all?

**Review by sriram-mv — APPROVED** [review]

File: tests/unit/lib/utils/test_osutils.py [hunk 2]
```diff
@@ -8,7 +8,7 @@
 from unittest import TestCase
 from unittest.mock import patch, Mock
 from samcli.lib.utils import osutils
-from samcli.lib.utils.osutils import rmtree_if_exists
+from samcli.lib.utils.osutils import rmtree_if_exists, rmtree_callback
```
Comment (lucashuy) [inline tests/unit/lib/utils/test_osutils.py:11]: Remove unused import

File: samcli/lib/utils/git_repo.py [hunk 3]
```diff
@@ -177,11 +177,22 @@ def _persist_local_repo(temp_path: str, dest_dir: Path, dest_name: str, replace_
...
+            msg = (
                 "Unstable state when updating repo. "
                 f"Check that you have permissions to create/delete files in {dest_dir} directory "
                 "or file an issue at https://github.com/aws/aws-sam-cli/issues"
-            ) from ex
+            )
+
+            if platform.system().lower() == "windows":
+                msg = (
+                    "Failed modify a local file when cloning app templates. "
```
Comment (hawflau) [inline samcli/lib/utils/git_repo.py:188]: nit:

```suggestion
                    "Failed to modify a local file when cloning app templates. "
```

File: tests/unit/lib/utils/test_git_repo.py [hunk 4]
```diff
@@ -249,3 +249,20 @@ def test_clone_with_longpaths_configured_in_windows(
...
+    @patch("samcli.lib.utils.git_repo.platform.system")
+    def test_clone_without_windows_longpath_exception_message(self, platform_mock, path_exist_mock):
+        path_exist_mock.side_effect = OSError()
+        platform_mock.return_value = "windows"
+
+        with self.assertRaises(CloneRepoUnstableStateException) as ex:
+            GitRepo._persist_local_repo(MagicMock(), MagicMock(), MagicMock(), MagicMock())
+
+        expected_msg = (
+            "Failed modify a local file when cloning app templates. "
```
Comment (hawflau) [inline tests/unit/lib/utils/test_git_repo.py:263]:
```suggestion
            "Failed to modify a local file when cloning app templates. "
```

**Review by hawflau — APPROVED** [review]
Approved with a tiny comment

**Pull request merged**